### ISSUE TITLE
Generic filter improvements

### ISSUE THREAD

**Issue by GlebDurygin** [Haulmont/jmix-ui#196]
The main issue Haulmont/jmix-ui#162

#### Filter actions
- [x] Implement `FilterMakeDefaultAction` class;
- [x] Implement `FilterSaveWithValuesAction` class;

#### Features
- [x] Add `parameterWhere`, `parameterJoin`, `parameterFetchPlan` attributes;
- [x] Jpql conditions;
- [x] Use filter with dynamic attributes;
- [x] Condition validation
- [x] Security

### XML
- [x] Add `<jpqlFilter/>` XML element;
- [x] Add `captionPosition` attribute to Filter component

#### Filter appearance
- [x] Remove unused CUBA styles from Filter component;
- [x] Use theme constants in screens;

### Events
- [x] `ConfigurationChangeEvent`

### Filter components
The generic filter component can only work with components that implement the `FilterComponent` interface. If the component is a container for placing multiple filter components, then it needs to implement `LogicalFilterComponent` interface.

![Untitled Diagram](https://user-images.githubusercontent.com/35256248/106612336-73ed0880-6582-11eb-98dc-38acf82103e1.png)

The filter component is treated as a query condition:
| FilterComponent |  Query condition  |
|:---------------:|:-----------------:|
|  PropertyFilter | PropertyCondition |
|    JpqlFilter   |   JpqlCondition   |
|   GroupFilter   |  LogicalCondition |

#### GroupFilter
The `GroupFilter` component is a composite component that has a `GroupBoxLayout` component as its root container.  This component is needed to combine several conditions into a logical group (AND by default, OR). The component can be used only inside the `Filter` component.

#### PropertyFilter
See Haulmont/jmix-ui#29, Haulmont/jmix-ui#171
 `PropertyFilter` is a UI component used for filtering entities returned by the {@link DataLoader}. The component is related to entity property and can automatically render proper layout for setting a condition value. In general case a PropertyFilter layout contains a label with entity property caption, operation label or selector (=, contains, >, etc.) and a field for editing a property value. The component can be used inside the `Filter` component and independently.
 ```
 <propertyFilter property="name"
                operation="CONTAINS"
                operationEditable="false">
    <textField id="nameFilterValueComponent"/>
</propertyFilter>
<propertyFilter property="age"
                operation="GREATER"
                operationEditable="true">
</propertyFilter>
 ```
 
 `propertyFilter` attributes inside the `filter`:
 - `id`
- `property` (required) May be a property or properties path: `name`, `order`, `order.date`
- `operation` (required)
- `operationEditable` Default **false**
- `operationCaptionVisible` Default **true**.
- `caption` (optional) If absent, the caption is evaluated by entity property
- `width`
- `parameterName` - name of LoadContext parameter that is used by this condition. If not defined then the parameter name is randomly generated.
- `defaultValue` - the default value of filter component

### JpqlFilter
`JpqlFilter` is a UI component used for filtering entities returned by the DataLoader. The component contains JPQL expressions that will be added to the 'from' and 'where' data loader query sections. The component can automatically render proper layout for setting a condition value. In general case a JpqlFilter layout contains a label with caption and a field for editing a condition value. The component can be used only inside the `Filter` component.

```
<jpqlFilter caption="Name like"
            parameterClass="java.lang.String"
            defaultValue="John">
    <condition>
        <c:jpql>
            <c:where>{E}.name like ?</c:where>
        </c:jpql>
    </condition>
</jpqlFilter>
```

 `jpqlFilter` attributes inside the `filter`:
 - `id`
- `parameterClass` (required) the value type
- `parameterName` - name of LoadContext parameter that is used by this condition. If not defined then the parameter name is randomly generated.
- `caption` (optional) If absent, the caption is evaluated by entity property
- `width`
- `defaultValue` - the default value of filter component

### Filter component registration
See `FilterComponentRegistration` class. The `FilterComponentRegistration` class has the following attributes:
* `componentClass` (required) - filter component class, must extend from `FilterComponent` class;
* `modelClass` (required) - non-persistent entity class that stores the state of the filter component, must extend from `FilterCondition` class. The model class is used to save filter component state in DB and is used to display and change the state of the filter component at runtime;
* `converterClass` (required) - a converter class for converting between a component and a model must implement the FilterConverter interface.
* `editorScreenId` (optional) - ID of the model edit screen. If no identifier is specified then the default identifier (`modelName.edit -> PropertyFilterCondition.edit`) will be used.
`PropertyFilter` registration example:
```
    @Bean
    public FilterComponentRegistration registerPropertyFilter() {
        return FilterComponentRegistrationBuilder.create(PropertyFilter.class,
                PropertyFilterCondition.class,
                PropertyFilterConverter.class)
                .withEditScreenId("ui_PropertyFilterCondition.edit")
                .build();
    }
```
All registered filter components are displayed in a `PopupButton` on the `AddConditionScreen` so that the user has the opportunity to create a filter component in runtime. Pressing the button will open the editor screen for the selected filter component.
<img width="318" alt="image" src="https://user-images.githubusercontent.com/35256248/101546072-abe03100-39c1-11eb-9106-1d026ccfaffd.png">

This method of registering filter components is convenient because we can register the filter components in different modules. And also the user can replace the filter component registered in the jmix framework with his own implementation by specifying the `@Order` annotation (for example, to expand the set of model attributes saved by the filter).

### Configurations
A configuration is a set of filter components that has an id (required) and name (optional). The id must be unique within this filter. If the name is not specified, then id is used as a key in the messagePack.
The `LogicalFilterComponent` is the root element of the configuration.
The configuration does not store a reference to all components, but stores a reference only to the root element (`LogicalFilterComponent`) from which the rest filter components can be obtained.
Configurations can be created at design time or at run time (via filter configuration editor).
Using Filter#getCurrentConfiguration() user can get the current configuration, and using Filter#getConfiguration(String id) user can get the configuration by id.

#### Design-time configuration
See `DesignTimeConfigurationImpl` class. Such configurations cannot be changed and saved in run-time, but the user can use `FilterCopyAction` that copies all conditions from the design-time configuration to run-time configuration.
`Configuration addConfiguration(String id, @Nullable String name)` method creates and adds to filter a design-time configuration.
Use `default` XML attribute to set the design-time configuration as default for filter.
All design-time configurations are displayed in the `selectConfigurationButton`:
<img width="318" alt="image" src="https://user-images.githubusercontent.com/35256248/101496487-3f464180-3983-11eb-8bee-aff7529dfe30.png">
```
<filter id="filter" dataLoader="testChildEntitiesDl">
    <properties include=".*"/>
    <configurations>
        <configuration id="test_configuration" name="Test Configuration 1 (AND)" operation="AND">
            <propertyFilter property="testName"
                            operation="ENDS_WITH"
                            operationEditable="false"/>
            <groupFilter operation="OR">
                <propertyFilter property="testInt"
                                operation="EQUAL"
                                operationCaptionVisible="false"/>
                <propertyFilter property="testDecimal"
                                operation="LESS_OR_EQUAL"/>
            </groupFilter>
        </configuration>
        <configuration id="test_configuration_2" name="Test Configuration 2 (OR)" operation="OR">
            <propertyFilter property="testName"
                            operation="ENDS_WITH"
                            operationEditable="false"/>
            <propertyFilter property="testDecimal"
                            operation="LESS_OR_EQUAL"/>
        </configuration>
    </configurations>
</filter>
```
#### Run-time configuration
See `RuntimeTimeConfigurationImpl`. Such configurations are created by the user in run-time and can be saved and modified.

### Predefined conditions
A predefined condition is a filter component that is not initially added to any of the configurations, but the user can select this component in the `AddConditionScreen` in the **Predefined conditions** section and add it to the run-time configuration.
```
<filter id="filter"
        dataLoader="testChildEntitiesDl">
    <properties include=".*"/>
    <conditions>
        <propertyFilter property="testTime"
                        operation="EQUAL"
                        operationEditable="true"
                        caption="Test time condition"/>
    </conditions>
</filter>
```

#### AddConditionScreen
The set of conditions that will be displayed in `AddConditionScreen` is defined using ConditionBuilder beans: PropertyConditionBuilder, PredefinedConditionBuilder, ConfigurationConditionBuilder.
![image](https://user-images.githubusercontent.com/35256248/106617273-e4e2ef00-6587-11eb-9737-3a4aa34cc9b5.png)

The set of possible dropdown actions for the Create button is determined by the number of registered FilterComponents.
![image](https://user-images.githubusercontent.com/35256248/106617636-3ee3b480-6588-11eb-8756-9f16f3a77a7d.png)

#### Configuration editor
![image](https://user-images.githubusercontent.com/35256248/106617908-72beda00-6588-11eb-8c09-f09a3da82044.png)

#### GroupCondition editor
![image](https://user-images.githubusercontent.com/35256248/106617729-51f68480-6588-11eb-9967-0b94177daf54.png)

#### PropertyCondition editor
![image](https://user-images.githubusercontent.com/35256248/106617799-5e7add00-6588-11eb-981c-5d0723bbcb40.png)

#### JpqlCondition editor
![image](https://user-images.githubusercontent.com/35256248/106618093-95e98980-6588-11eb-9a5e-93d65756857e.png)

### Filter actions
Now, the `Filter` component implements `ActionsHolder` interface. All actions will be added to the `settingsButton` PopupButton.
<img height="250" alt="image" src="https://user-images.githubusercontent.com/35256248/101454734-e1960300-394a-11eb-9a33-c517e64bdef9.png">

The framework loads a set of default actions for the filter, depending on the connected modules:
* ui module
  * `FilterEditAction` (**filter_edit**)
  * `FilterCopyAction` (**filter_copy**)
  * `FilterClearValuesAction` (**filter_clearValues**)
* ui-data module
  * `FilterSaveAction` (**filter_save**)
  * `FilterSaveWithValuesAction` (**filter_saveWithValues**)
  * `FilterSaveAsAction` (**filter_saveAs**)
  * `FilterEditAction` (**filter_edit**)
  * `FilterRemoveAction` (**filter_remove**)
  * `FilterCopyAction` (**filter_copy**)
  * `FilterMakeDefaultAction` (**filter_makeDefault**)
  * `FilterClearValuesAction` (**filter_clearValues**)
  
  User can override default actions in XML or in JAVA:
```
<filter id="filter1" dataLoader="testChildEntitiesDl">
    <properties include=".*"/>
    <actions>
        <action id="addCondition" type="filter_addCondition"/>
        <action id="edit" type="filter_edit"/>
    </actions>
</filter>
```
```
filter.removeAllActions();
FilterAction filterAction = actions.create(FilterAddCondition.class);
filterAction.setFilter(filter);
filter.addAction(filterAction);
```

**filter_save** - saves changes to current filter configuration.
**filter_saveWithValues** - saves changes to the current filter configuration using the values in inner filter components as filter default values.
**filter_saveAs** - saves current filter configuration under new code and caption;
**filter_edit** - edits current run-time filter configuration (opens `FilterConfigurationEdit` screen). Disabled for design-time configurations.
**filter_remove** - removes current run-time filter configuration. Disabled for design-time configurations.
**filter_copy** - copies all conditions from design-time configuration to run-time configuration;
**filter_makeDefault** - makes the filter configuration default for this screen. The filter will be automatically displayed on the filter panel when the screen is opened. Takes precedence over `default` attribute of design-time configurations and `Default for all users` attribute of configuration. Disabled if no `screenSettings` facet is added to the screen.
**filter_clearValues** - clears the filter condition values.
**filter_addCondition** - adds condition to current filter configuration.

### Permissions
- **ui.filter.modifyGlobalConfiguration** - to create/change/delete global (available to all users) filter configurations
- **ui.filter.modifyJpqlCondition** - to create/change jpql conditions in run-time

### Ui properties
- **jmix.ui.filter.autoApply** - allows controlling the default search mode for the filter. `autoApply` attribute specifies when the filter is applied. If the attribute value is false, the filter will be applied only after the Search button is clicked. If the attribute value is true, filter is applied immediately after changing the filter components.
- **jmix.ui.filter.propertiesHierarchyDepth** - defines the properties hierarchy depth in the "AddConditionScreen"
- **jmix.ui.filter.columnsCount** - sets the default number of columns for placing FilterComponents on the filter panel
- **jmix.ui.filter.showConfigurationIdField** - determines whether the configurationIdField should be shown in editors. By default, the configurationIdField is not shown and the configurationId is generated depending on the configurationName.

##### Add a design-time configuration with a PropertyFilter to a Filter
* XML
    ```
    <filter id="filter" dataLoader="ordersDl">
        <properties include=".*"/>
        <configurations>
            <configuration id="defaultXmlConfiguration" name="msg://defaultXmlConfiguration.name" default="true">
                <propertyFilter property="customer" operation="EQUAL"
                                operationEditable="true"/>
                <propertyFilter property="customer.grade" operation="EQUAL"
                                operationEditable="true"/>
            </configuration>
        </configurations>
    </filter>
    ```
* Java
    ```
    @Autowired
    protected UiComponents uiComponents;
    @Autowired
    protected PropertyFilterSupport propertyFilterSupport;
    @Autowired
    protected SingleFilterSupport singleFilterSupport;
    @Autowired
    protected Filter filter;

    @Autowired
    protected Messages messages;

    @Subscribe
    protected void onInit(InitEvent event) {
        DesignTimeConfiguration javaDefaultConfiguration = filter.addConfiguration("javaDefaultConfiguration",
                "Default configuration");
        DataLoader dataLoader = filter.getDataLoader();

        PropertyFilter<Customer> customerPropertyFilter = uiComponents.create(PropertyFilter.NAME);
        customerPropertyFilter.setDataLoader(dataLoader);
        customerPropertyFilter.setProperty("customer");
        customerPropertyFilter.setOperation(PropertyFilter.Operation.EQUAL);
        customerPropertyFilter.setOperationEditable(true);
        customerPropertyFilter.setCaption(propertyFilterSupport.getPropertyFilterCaption(
                dataLoader.getContainer().getEntityMetaClass(),
                customerPropertyFilter.getProperty()));
        customerPropertyFilter.setParameterName(PropertyConditionUtils.generateParameterName(
                customerPropertyFilter.getProperty()));
        customerPropertyFilter.setValueComponent(singleFilterSupport.generateValueComponent(
                dataLoader.getContainer().getEntityMetaClass(),
                customerPropertyFilter.getProperty(),
                customerPropertyFilter.getOperation()
        ));
        javaDefaultConfiguration.getRootLogicalFilterComponent().add(customerPropertyFilter);

        PropertyFilter<CustomerGrade> gradePropertyFilter = uiComponents.create(PropertyFilter.NAME);
        gradePropertyFilter.setDataLoader(dataLoader);
        gradePropertyFilter.setProperty("customer.grade");
        gradePropertyFilter.setOperation(PropertyFilter.Operation.EQUAL);
        gradePropertyFilter.setOperationEditable(true);
        gradePropertyFilter.setCaption(propertyFilterSupport.getPropertyFilterCaption(
                dataLoader.getContainer().getEntityMetaClass(),
                gradePropertyFilter.getProperty()));
        gradePropertyFilter.setParameterName(PropertyConditionUtils.generateParameterName(
                gradePropertyFilter.getProperty()));
        gradePropertyFilter.setValueComponent(singleFilterSupport.generateValueComponent(
                dataLoader.getContainer().getEntityMetaClass(),
                gradePropertyFilter.getProperty(),
                gradePropertyFilter.getOperation()
        ));
        javaDefaultConfiguration.getRootLogicalFilterComponent().add(gradePropertyFilter);
        filter.setCurrentConfiguration(javaDefaultConfiguration);        
    }
    ```

##### Add a design-time configuration with a JpqlFilter to a Filter
* XML
    ```
    <filter id="filter" dataLoader="ordersDl">
        <properties include=".*"/>
        <configurations>
            <configuration id="defaultXmlConfiguration" name="msg://defaultXmlConfiguration.name" default="true">
                <jpqlFilter caption="msg://jpqlFilter.caption"
                            parameterClass="io.jmix.sampler.entity.Product">
                    <condition>
                        <c:jpql>
                            <c:join>join {E}.items i</c:join>
                            <c:where>i.product.id = ?</c:where>
                        </c:jpql>
                    </condition>
                </jpqlFilter>
            </configuration>
        </configurations>
    </filter>
    ```
* Java
    ```
    @Autowired
    protected UiComponents uiComponents;
    @Autowired
    protected JpqlFilterSupport jpqlFilterSupport;
    @Autowired
    protected SingleFilterSupport singleFilterSupport;
    @Autowired
    protected Filter filter;

    @Autowired
    protected Messages messages;

    @Subscribe
    protected void onInit(InitEvent event) {
        DesignTimeConfiguration javaDefaultConfiguration = filter.addConfiguration("javaDefaultConfiguration",
                "Default configuration");
        DataLoader dataLoader = filter.getDataLoader();

        JpqlFilter<Product> jpqlFilter = uiComponents.create(JpqlFilter.NAME);
        jpqlFilter.setFrame(getWindow());
        jpqlFilter.setDataLoader(dataLoader);
        jpqlFilter.setCondition("i.product.id = ?", "join {E}.items i");
        jpqlFilter.setParameterClass(Product.class);
        jpqlFilter.setCaption(messages.getMessage(FilterJpqlConditionSample.class, "jpqlFilter.caption"));
        jpqlFilter.setParameterName(jpqlFilterSupport.generateParameterName(
                jpqlFilter.getId(),
                jpqlFilter.getParameterClass().getSimpleName()));
        jpqlFilter.setValueComponent(singleFilterSupport.generateValueComponent(
                dataLoader.getContainer().getEntityMetaClass(),
                jpqlFilter.hasInExpression(),
                jpqlFilter.getParameterClass()
        ));

        javaDefaultConfiguration.getRootLogicalFilterComponent().add(jpqlFilter);
        filter.setCurrentConfiguration(javaDefaultConfiguration);    
    }
    ```
##### Add a design-time configuration with a GroupFilter to a Filter
* XML
    ```
    <filter id="filter" dataLoader="customersDl">
        <properties include=".*"/>
        <configurations>
            <configuration id="defaultXmlConfiguration" name="msg://defaultXmlConfiguration.name" default="true">
                <groupFilter operation="OR">
                    <propertyFilter property="active" operation="EQUAL"
                                    operationEditable="true" defaultValue="false"/>
                    <propertyFilter property="grade" operation="EQUAL"
                                    operationEditable="true" defaultValue="30"/>
                </groupFilter>
                <propertyFilter property="age" operation="GREATER_OR_EQUAL"
                                operationEditable="true" defaultValue="30"/>
            </configuration>
        </configurations>
    </filter>
    ```
* Java
    ```
    @Autowired
    protected UiComponents uiComponents;
    @Autowired
    protected PropertyFilterSupport propertyFilterSupport;
    @Autowired
    protected SingleFilterSupport singleFilterSupport;
    @Autowired
    protected LogicalFilterSupport logicalFilterSupport;
    @Autowired
    protected Filter filter;

    @Autowired
    protected Messages messages;

    @Subscribe
    protected void onInit(InitEvent event) {
        DesignTimeConfiguration javaDefaultConfiguration = filter.addConfiguration("javaDefaultConfiguration",
                "Default configuration");
        DataLoader dataLoader = filter.getDataLoader();

        GroupFilter groupFilter = uiComponents.create(GroupFilter.NAME);
        groupFilter.setDataLoader(dataLoader);
        groupFilter.setOperation(LogicalFilterComponent.Operation.OR);
        groupFilter.setCaption(logicalFilterSupport.getOperationCaption(groupFilter.getOperation()));

        PropertyFilter<Boolean> activePropertyFilter = uiComponents.create(PropertyFilter.NAME);
        activePropertyFilter.setDataLoader(dataLoader);
        activePropertyFilter.setProperty("active");
        activePropertyFilter.setOperation(PropertyFilter.Operation.EQUAL);
        activePropertyFilter.setOperationEditable(true);
        activePropertyFilter.setCaption(propertyFilterSupport.getPropertyFilterCaption(
                dataLoader.getContainer().getEntityMetaClass(),
                activePropertyFilter.getProperty()));
        activePropertyFilter.setParameterName(PropertyConditionUtils.generateParameterName(
                activePropertyFilter.getProperty()));
        activePropertyFilter.setValueComponent(singleFilterSupport.generateValueComponent(
                dataLoader.getContainer().getEntityMetaClass(),
                activePropertyFilter.getProperty(),
                activePropertyFilter.getOperation()
        ));
        groupFilter.add(activePropertyFilter);

        PropertyFilter<CustomerGrade> gradePropertyFilter = uiComponents.create(PropertyFilter.NAME);
        gradePropertyFilter.setDataLoader(dataLoader);
        gradePropertyFilter.setProperty("grade");
        gradePropertyFilter.setOperation(PropertyFilter.Operation.EQUAL);
        gradePropertyFilter.setOperationEditable(true);
        gradePropertyFilter.setCaption(propertyFilterSupport.getPropertyFilterCaption(
                dataLoader.getContainer().getEntityMetaClass(),
                gradePropertyFilter.getProperty()));
        gradePropertyFilter.setParameterName(PropertyConditionUtils.generateParameterName(
                gradePropertyFilter.getProperty()));
        gradePropertyFilter.setValueComponent(singleFilterSupport.generateValueComponent(
                dataLoader.getContainer().getEntityMetaClass(),
                gradePropertyFilter.getProperty(),
                gradePropertyFilter.getOperation()
        ));
        groupFilter.add(gradePropertyFilter);
        javaDefaultConfiguration.getRootLogicalFilterComponent().add(groupFilter);

        PropertyFilter<Integer> agePropertyFilter = uiComponents.create(PropertyFilter.NAME);
        agePropertyFilter.setDataLoader(dataLoader);
        agePropertyFilter.setProperty("age");
        agePropertyFilter.setOperation(PropertyFilter.Operation.GREATER_OR_EQUAL);
        agePropertyFilter.setOperationEditable(true);
        agePropertyFilter.setCaption(propertyFilterSupport.getPropertyFilterCaption(
                dataLoader.getContainer().getEntityMetaClass(),
                agePropertyFilter.getProperty()));
        agePropertyFilter.setParameterName(PropertyConditionUtils.generateParameterName(
                agePropertyFilter.getProperty()));
        agePropertyFilter.setValueComponent(singleFilterSupport.generateValueComponent(
                dataLoader.getContainer().getEntityMetaClass(),
                agePropertyFilter.getProperty(),
                agePropertyFilter.getOperation()
        ));

        javaDefaultConfiguration.getRootLogicalFilterComponent().add(agePropertyFilter);

        // Set default values for configuration
        activePropertyFilter.setValue(false);
        javaDefaultConfiguration.setFilterComponentDefaultValue(activePropertyFilter.getParameterName(), false);

        gradePropertyFilter.setValue(CustomerGrade.STANDARD);
        javaDefaultConfiguration.setFilterComponentDefaultValue(gradePropertyFilter.getParameterName(), CustomerGrade.STANDARD);

        agePropertyFilter.setValue(30);
        javaDefaultConfiguration.setFilterComponentDefaultValue(agePropertyFilter.getParameterName(), 30);
    }
    ```
##### Add a filter component to existing configuration
```
    @Autowired
    protected UiComponents uiComponents;
    @Autowired
    protected PropertyFilterSupport propertyFilterSupport;
    @Autowired
    protected SingleFilterSupport singleFilterSupport;

    @Autowired
    protected Filter filter;

    @Subscribe
    protected void onAfterInit(AfterInitEvent event) {
        DataLoader dataLoader = filter.getDataLoader();

        PropertyFilter<Integer> agePropertyFilter = uiComponents.create(PropertyFilter.NAME);
        agePropertyFilter.setDataLoader(dataLoader);
        agePropertyFilter.setProperty("age");
        agePropertyFilter.setOperation(PropertyFilter.Operation.GREATER_OR_EQUAL);
        agePropertyFilter.setOperationEditable(true);
        agePropertyFilter.setCaption(propertyFilterSupport.getPropertyFilterCaption(
                dataLoader.getContainer().getEntityMetaClass(),
                agePropertyFilter.getProperty()));
        agePropertyFilter.setParameterName(PropertyConditionUtils.generateParameterName(
                agePropertyFilter.getProperty()));
        agePropertyFilter.setValueComponent(singleFilterSupport.generateValueComponent(
                dataLoader.getContainer().getEntityMetaClass(),
                agePropertyFilter.getProperty(),
                agePropertyFilter.getOperation()
        ));

        filter.getCurrentConfiguration().getRootLogicalFilterComponent().add(agePropertyFilter);
        filter.refreshCurrentConfigurationLayout();
    }
```
    
##### Programmatically defined Filter
```
    @Autowired
    protected UiComponents uiComponents;
    @Autowired
    protected PropertyFilterSupport propertyFilterSupport;
    @Autowired
    protected SingleFilterSupport singleFilterSupport;
    @Autowired
    protected Messages messages;

    @Autowired
    protected CollectionLoader<Customer> customersDl;

    @Subscribe
    protected void onAfterInit(AfterInitEvent event) {
        Filter filter = uiComponents.create(Filter.NAME);
        filter.setId("filter");
        getWindow().add(filter, 0);
        filter.setDataLoader(customersDl);

        filter.loadConfigurationsAndApplyDefault();

        DesignTimeConfiguration javaDefaultConfiguration = filter.addConfiguration("javaDefaultConfiguration",
                "Default configuration");

        PropertyFilter<Integer> agePropertyFilter = uiComponents.create(PropertyFilter.NAME);
        agePropertyFilter.setDataLoader(customersDl);
        agePropertyFilter.setProperty("age");
        agePropertyFilter.setOperation(PropertyFilter.Operation.LESS_OR_EQUAL);
        agePropertyFilter.setOperationEditable(true);
        agePropertyFilter.setCaption(propertyFilterSupport.getPropertyFilterCaption(
                customersDl.getContainer().getEntityMetaClass(),
                agePropertyFilter.getProperty()));
        agePropertyFilter.setParameterName(PropertyConditionUtils.generateParameterName(
                agePropertyFilter.getProperty()));
        agePropertyFilter.setValueComponent(singleFilterSupport.generateValueComponent(
                customersDl.getContainer().getEntityMetaClass(),
                agePropertyFilter.getProperty(),
                agePropertyFilter.getOperation()
        ));

        javaDefaultConfiguration.getRootLogicalFilterComponent().add(agePropertyFilter);
        filter.setCurrentConfiguration(javaDefaultConfiguration);
    }
```

**Comment by maistrenkoIulia** [comment]
TC:
ser for browser filter with actions:
```
<filter dataLoader="personsDl">
    <properties include=".*"/>
    <actions>
        <action id="1" type="filter_addCondition"/>
        <action id="2" type="filter_clearValues"/>
        <action id="3" type="filter_copy"/>
        <action id="4" type="filter_edit"/>
        <action id="5" type="filter_makeDefault"/>
        <action id="6" type="filter_remove"/>
        <action id="7" type="filter_save"/>
        <action id="8" type="filter_saveAs"/>
        <action id="9" type="filter_saveWithValues"/>
    </actions>
</filter>
```
run app>open browser>filter_settingsButton
ER: save as, coby actions should be available
AR: all actions available

**Comment by maistrenkoIulia** [comment]
captionPosition attr isn`t find by idea
![image](https://user-images.githubusercontent.com/66992566/107485503-49600880-6b9d-11eb-9a64-4f54b3b71c95.png)

**Comment by maistrenkoIulia** [comment]
TC: set jmix.ui.showFilterConfigurationIdField =true
run app
edit filter
ER: confiuration id should be aligned left
![image](https://user-images.githubusercontent.com/66992566/107624616-96fb7480-6c74-11eb-8c59-e0f8d9bcb5c5.png)

**Comment by maistrenkoIulia** [comment]
missing confirmation popup during filter>remove

**Comment by maistrenkoIulia** [comment]
need to localize hint for Parameter Name in 'Proprty Condition Editor'
![image](https://user-images.githubusercontent.com/66992566/109021766-9fbc6380-76d4-11eb-9bbf-1e9384e2c48c.png)

**Comment by GlebDurygin** [comment]
Will be fixed in Haulmont/jmix-ui#343

> need to localize hint for Parameter Name in 'Proprty Condition Editor'
